designs/2018-programmatic-style-edits/README.md

```diff
@@ -0,0 +1,362 @@
+
+**Important:** Style editors are not applied when the `--fix` and `--fix-dry-run` flags are not used.
+
+### Error Reporting
+
+If a style editor attempts to use its own parser to parse text, and the parser throws an error, ESLint will report the error in the same format as a parser error thrown during linting.
+
+If a style editor only modifies text, and the modified text throws an error while the linter is parsing it, the same error format is displayed as it would with any source code that hasn't been style fixed.
+
+### Using with `--fix-type`
+
+In order to allow control over when style editors will be applied, a fourth option will be added to `--fix-type` called `style`. When `style` is specified using `--fix-type`, any specified style editors will be applied.
+
+To apply only the style editor, an end-user would run:
+
+```
+$ eslint --fix --fix-type style *.js
+```
+
+To apply just layout rules and the style editor, an end-user would run:
+
+```
+$ eslint --fix --fix-type layout,style *.js
+```
+
+This would *not* run a style editor and only fix problems:
+
+```
+$ eslint --fix --fix-type problem *.js
+```
+
+### Implementation Details
+
+I envision the implementation as follows:
+
+1. Update `CLIEngine` and `cli` to accept `style` as a valid option for `--fix-type`.
+1. Create a `Style` class to represent individual style editors.
+1. When plugins are loaded, look for style editors in addition to rules and processors.
+1. Style editors would be loaded and applied from `CLIEngine`.
+1. Add a `Linter#defineStyle()` method to store style editors inside of `Linter`.
+
+## Documentation
+
+This feature would need a new documentation section to explain how to create style editors. Additionally, the following pages would need to be updated:
+
+* [Configuring ESLint](https://eslint.org/docs/user-guide/configuring#configuring-eslint)
+* [Working with Plugins](https://eslint.org/docs/developer-guide/working-with-plugins#working-with-plugins)
+* [Command Line Interface](https://eslint.org/docs/user-guide/command-line-interface#command-line-interface)
+
+## Drawbacks
+
+* This proposal adds new functionality that will have to be taught to existing users.
+* It will take some time for the community to create style editors and therefore adoption might be slow.
+* Applying style guides does add some complexity to the linting process that doesn't already exist.
+* Adding two additional steps to the autofixing process will slow down the process as a whole by some unknown amount.
+* Style editors do not allow for programmatic detection of whether or not the style has been applied.
+
+## Backwards Compatibility Analysis
+
+Because style editors are a new feature built on top of existing features, backwards compatibility is not a concern.
+
+If people don't specify a style editor, then their experience doesn't change at all.
+
+## Alternatives
```

Comment:
Another alternative would be to have users use separate formatting tools after running ESLint (e.g. by having a script that runs ESLint and then formats their code with `prettier`).

This would avoid the need to add complexity to ESLint core, but as a downside it wouldn't work as well with ESLint processors.